src/Elastic.Apm.AspNetCore/ApmMiddleware.cs

```diff
@@ -144,6 +146,24 @@ private void FillSampledTransactionContextRequest(HttpContext context, Transacti
 					requestHeaders.Add(header.Key, header.Value.ToString());
 			}
 
+			var body = Consts.BodyRedacted; // According to the documentation - the default value of 'body' is '[Redacted]'
+			//If 'CaptureBody' is not 'off' and 
+			if (hasBody(context.Request) &&
+				(_configurationReader.CaptureBody.Equals(ConfigConsts.SupportedValues.CaptureBodyAll) ||
+				_configurationReader.CaptureBody.Equals(ConfigConsts.SupportedValues.CaptureBodyTransactions))
+				)
+			{
+				using (var reader = new StreamReader(context.Request.Body))
+				{
+					body = reader.ReadToEnd();

```

Comment:
I see some exceptions here when I run the sample app:

```
fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[1]
      An unhandled exception has occurred while executing the request.
System.IO.IOException: Unexpected end of Stream, the content may have already been read by another component. 
   at Microsoft.AspNetCore.WebUtilities.MultipartReaderStream.ReadAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.WebUtilities.StreamHelperExtensions.DrainAsync(Stream stream, ArrayPool`1 bytePool, Nullable`1 limit, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.WebUtilities.MultipartReader.ReadNextSectionAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Http.Features.FormFeature.InnerReadFormAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Mvc.ModelBinding.FormValueProviderFactory.AddValueProviderAsync(ValueProviderFactoryContext context)
   at Microsoft.AspNetCore.Mvc.ModelBinding.CompositeValueProvider.CreateAsync(ActionContext actionContext, IList`1 factories)
   at Microsoft.AspNetCore.Mvc.ModelBinding.CompositeValueProvider.CreateAsync(ControllerContext controllerContext)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<<CreateBinderDelegate>g__Bind|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
   at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)

```

I googled this and it seems to be a common thing. [This one](https://stackoverflow.com/questions/49867343/unexpected-end-of-stream-the-content-may-have-already-been-read-by-another-comp) suggest disabling model binding, but that does not help in our case - we can't tell this to all our users.

I'm trying to help here and come up with something. I'd say this is definitely something where we should come up with a solution for.

Reply:
Looking into this as well.

Reply:
@gregkalapos - I wasn't able to reproduce this problem (please provide instructions on who to do it). However - the problem of an "already read" request/response stream is a well known one. There are several approaches to resolve this , with the one most appropriate for this problem being 
1. Resetting the body stream reader to position 0 when it's supported.
2. Reading the body and then writing it back to the memory to prevent subsequent attempts to the read the body from failing. 
I'm implementing and testing it now.

There is another other solution which I think are not suitable for this case-
- EnableRewind() for all requests in the pipeline but that would mean we are changing the behavior of the web app on the server. 

refs : 
https://gunnarpeipman.com/aspnet/aspnet-core-request-body/
http://www.palador.com/2017/05/24/logging-the-body-of-http-request-and-response-in-asp-net-core/
https://devblogs.microsoft.com/aspnet/re-reading-asp-net-core-request-bodies-with-enablebuffering/

Reply:
Code committed. Please check on your end if possible or provide instructions for recreating this problem.

Reply:
How do you test? I simply sent an HTTP POST to `http://localhost:5000/api/Home/Post` with Postman. If I remember correctly, HTTP GET requests from a browser did not repro the problem for me either (I should have said this earlier.)

To the current version: I still see some problems, and this time it happens always:
```
Application started. Press Ctrl+C to shut down.
fail: Elastic.Apm[0]
      {ApmMiddleware} General Error reading request body , Error msg : Specified method is not supported. , stack trace :    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.set_Position(Int64 value)
         at Elastic.Apm.AspNetCore.Extensions.HttpRequestExtensions.ExtractRequestBody(HttpRequest request, IApmLogger logger) in /Users/gergelykalapos/Repos/apm-agent-dotnet/src/Elastic.Apm.AspNetCore/Extensions/RequestExtentions.cs:line 28 
fail: Elastic.Apm[0]
      {ApmMiddleware} General Error reading request body , Error msg : Specified method is not supported. , stack trace :    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.set_Position(Int64 value)
         at Elastic.Apm.AspNetCore.Extensions.HttpRequestExtensions.ExtractRequestBody(HttpRequest request, IApmLogger logger) in /Users/gergelykalapos/Repos/apm-agent-dotnet/src/Elastic.Apm.AspNetCore/Extensions/RequestExtentions.cs:line 28 
^CApplication is shutting down...
```
It's very strange that [on this post](https://gunnarpeipman.com/aspnet/aspnet-core-request-body/) no one mentions this problem.

Doesn't this happen on your machine?

This GitHub issue is the same: https://github.com/aspnet/KestrelHttpServer/issues/1548 

And the solution for that is to call `request.EnableRewind();` - that's currently not available, for that we need a reference to the `Microsoft.AspNetCore.Http` package. But it's still very strange that on the 1. blogpost people seem to suggest that as an alternative, and no one says that the original version without `request.EnableRewind();`  does not seem to work. It defiantly fails on my machine and [the referenced GitHub issue](https://github.com/aspnet/KestrelHttpServer/issues/1548 ) seems to verify that without `request.EnableRewind();` it should not work.

Reply:
@gregkalapos - I tested using the SampleAspNetCoreApp and with test methods (
TestJsonBodyRetrievalOnRequestFailureInAspNetCore, FailingPostRequestWithoutConfiguredExceptionPage, HomeSimplePagePostTransactionTest, all in AspNetCoreMiddlewareTests.cs).
I'll try it your way. The test methods for this worked but since I merged master into my working branch - these tests fail. It looks like the initialization of the agent in the test method (with the captureBody setting being set to all/errors/transactions) is being ignored and then debugging I can see that the loaded value in the website being used in the test is "off". I'm still trying to understand why this is happening.

Reply:
I removed the request.body.position=0 statement after re-creating the problem.
Still working on the 'Unexpected end of Stream' problem (Which I haven't been able to recreate yet).